strawberryfields/engine.py

```diff
@@ -308,7 +308,8 @@ def _broadcast_nones(val, dim):
             self.run_progs.append(p) 
             prev = p
 
-        return Result(self.samples.copy())
+        if self.samples:
+            return Result(self.samples.copy())
```

Comment:
What is the reason behind this change here? The `Result` class stores both the quantum state as well as the samples for the `LocalEngine`, so likely still needs to be returned even if no samples are requested

Reply:
In the case when a user aborts the engine (e.g. with KeyboardInterrupt and the job goes to the background / is async), `self.samples` is `None` and this fails. I did not recognize the use case you mentioned with LocalEngine (I assume samples in that case could be an empty list?) I can change this to `if self.samples is not None` to resolve this issue.

NB: In the case of a KeyboardInterrupt/async, the current behaviour requires the user to continue or pick up the job manually (i.e. reload/fetch results, and wrap them in a `Result` object) but in the future we can add a helper method to resume the loop.

Reply:
With `LocalEngine`, samples will always be a list of size the `(shots, num_modes)` (iirc) if no shots are requested. Each value in the list will simply be `None`.

This line creates the `Result` object (even if the samples are all `None`), and later on the `State` object is attached to `result.state`. See https://github.com/XanaduAI/strawberryfields/blob/14f91bf5f2e0dc60b2a431c23c8edf6ce5dd1850/strawberryfields/engine.py#L397.

My understanding is that this line will raise an exception if `result` is now `None`.

Reply:
Ah, this is actually very subtle. Because the simulator backends return a list of the form `[None, None, None]` if there is no sampling, the `if self.samples:` check passes.

I think this is fine for now, but we should leave a comment here explaining exactly what is happening, so that when we return to finish #111 there isn't any confusion/bugs